### ISSUE TITLE
Proposal: Specify columns to ignore during reverse engineering

### ISSUE THREAD

**Issue by sm15455** [ErikEJ/EFCorePowerTools#572]
Hi,

when scaffolding from database, I often need to map a property to an enum type and not to its database column type. I know I can use handlebars template, but I have another suggestion which could make life easier. Would it be possible to expand the renaming file (or create a new one) to specify columns to ignore during import? This would allow me to use partial classes to define properties for those columns and partial method in OnModelCreating to define their mapping. I know such task would not be high in the priority list, so I would be more than happy to contribute.

**Comment by ErikEJ** [comment]
Please give an example of your modifications if a column was not included.

**Comment by sm15455** [comment]
`
//EFPT generated code
public partial class Person
{
  public int Id { get; set; }
  public string Name { get; set; }
}

//my custom code
public partial class Person
{
  public Gender Gender { get; set; }
}

public enum Gender 
{
  Male,
  Female,
  Other
}

public partial class MyDbContext
{
  void OnModelCreatingPartial(ModelBuilder modelBuilder)
  {
    modelBuilder.Entity<Person>(entity => 
    {
      entity.Property(e => e.Gender)
        .whatevermapping();
    });
  }
}
`

**Comment by ErikEJ** [comment]
Had a closer look at this, looks like an approach to try would be to implement https://github.com/dotnet/efcore/blob/main/src/EFCore.Design/Scaffolding/Internal/RelationalScaffoldingModelFactory.cs#L299 and register this override when building the service provider, if a list of columns to exclude are available, then modify the DatabaseTable model by removing the excluded columns.

we would need a way to pass the list of columns to exclude to this class. 

The list of columns to exclude could be under each Table in the tables list in efpt.config.json.

PRs accepted.

**Comment by sm15455** [comment]
That's the file I was looking at. I'm glad we have same idea. 
I cloned the repository and will start working on a PR as soon as I can. The only problem is that I never created a VS Extension before so it will take some time before everything is setup.

**Comment by ErikEJ** [comment]
@sm15455 As long as you have the Extensibility workload installed in Visual Studio, set EFCorePowerTools project as startup and set build platform to x86, it should be as simple a pressing F5! 

Please feel free to discuss any design concerns before getting in too deep.

**Comment by ErikEJ** [comment]
For testing just scaffolding, make the efreveng project your startup project!

**Comment by ErikEJ** [comment]
Duplicate of #353

**Comment by paul1923** [comment]
Another use case when ignoring column names is helpful is Temporal Tables, ie:  [SysEndTime], [SysStartTime] columns

**Comment by ErikEJ** [comment]
@paul1923 Sure, but you can also mark them hidden.

**Comment by sm15455** [comment]
@ErikEJ I'm slowly moving on. 

On The reveng side I did the following changes
- Modified the TableInformationModel class adding the ExcludedColumns (string[]) property.
- Inherited from RelationalScaffoldingModelFactory and overriden the VisitTable method to remove excluded columns before the visitor creates the entity builder for the table.
- Added the new scaffolder to the service provider builder.

Now I need to work on the UI side. I should:
- Serialize and deserialize the ExcludedColumns property as a property of a table in the eft.config.json.
- create a hierarchical view of the tables where user can view all tables and their columns. All columns for a table should be selected and the user must select those he wants to exclude.
- Remember the user selection when UI is reloaded (it happens now for table selection and must do the same for columns)

The problem is that TableInformationModel is directly used in the ITableInformationViewModel model so I can't use the ExcludedColumns property because in the UI it's not a simple list of strings, but a list of (string, bool) (needed for selection). The most conservative approach I can think of is to add a ExcludedColumns property in the ITableInformationViewModel and then write code to map it to and from ExcludedColumns property in the TableInformationModel. Any thought on this?

**Comment by ErikEJ** [comment]
Fantastic!

I suggest that we make it a manual feature for now, so folks can try it out, just make sure that we serialize and deserialize properly.

Then we can also think about improving the View Model, possibly map to a class that is better suited for UI display than the TableInformationModel.

**Comment by ErikEJ** [comment]
Submitting a PR with that will also make the changes and code review much more manageable.

**Comment by paul1923** [comment]
@ErikEJ 

I get around this by modifying the DbContext with Ignore entries, ie...
```
   entity.Ignore(x => x.SysEndTime);
   entity.Ignore(x => x.SysStartTime);
```
I need to do this for each DbSet.  

I wasn't aware there was a hidden option, how do I access that

**Comment by ErikEJ** [comment]
@paul1923 HIDDEN is part of the CREATE table statement

**Comment by sm15455** [comment]
@ErikEJ I need permission to push a new branch.

**Comment by ErikEJ** [comment]
You need to fork the project to your own Github repo, then create a Pull Request - are you not able to do that?

**Comment by sm15455** [comment]
Sorry, my first contribution on a GitHub project. I thought I should push a branch to this repo :D

**Comment by ErikEJ** [comment]
No worries, been there myself - let me know if you are unblocked.

**Comment by paul1923** [comment]
@ErikEJ 

Thanks for the tip on HIDDEN columns.  I've opted instead to create another DbContext partial class for the fields I want ignored.

**Comment by sm15455** [comment]
@ErikEJ I close this issue as the feature is now implemented. I'll open a new one for the UI

**Comment by ErikEJ** [comment]
It is not in the daily build yet, hang in there...

**Comment by ErikEJ** [comment]
@sm15455 

Some thoughts on next step:

Create new class to exchange the table list (stop using TableInformation).
Add columns to this class.

Add columns to the output from TableList command if EFPT CLI tool.

Extend TableInformation with a new, non-serialized List<ColumnInformation>, where ColumnInformation has:

string Name
bool Excluded

Construct the ExcludeColumns property when loading options, and set when saving options.

Bind ColumnInfomation to UI - no idea how...

**Comment by ErikEJ** [comment]
Custom [enum mapping](https://github.com/ErikEJ/EFCorePowerTools/issues/1472#issuecomment-1220273164) is now available with EF Core 7 and T4 templates

**Comment by Colemartindale** [comment]
> @paul1923 Sure, but you can also mark them hidden.

How do you mark Temporal tables as hidden?

**Comment by Colemartindale** [comment]
My apologies.. I'm not following. Should this be done on the DB side or in the app itself?
We recently updated our API to .NET & EF 6.0. Within our scaffolded DbContext file, the temporal tables now get included (when they didn't before) on our model builder.

We have a context extension class that implements this method:

`partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
    {
      modelBuilder.Entity<AnalysisMetric>(entity => 
                     {entity.Ignore("analysis_metric_HISTORY");
            });
    }`

but this doesn't seem to work

**Comment by ErikEJ** [comment]
Please create a new issue and share a repro script and I will have a look.
